File: packages/engine/src/machine.ts
import { areAdjacent, getTile, swapTiles } from './board.js';
import { evaluateChain } from './equation.js';
import { createInitialState } from './generator.js';
import { eq } from './num.js';
import { refillAfterRemoval } from './refill.js';
import { analyseWithBand } from './solver.js';
import type { GameAction, LevelState, Phase } from './types.js';
import { validatePuzzle } from './validator.js';

const INPUT_PHASES: readonly Phase[] = ['ready', 'dragging', 'previewing'];

function appendHistory(state: LevelState, action: GameAction): readonly GameAction[] {
  return [...state.history, action];
}

function objectiveComplete(state: LevelState, score: number, solvedCount: number): boolean {
  return state.rules.objective === 'scoreTide'
    ? score >= state.rules.goalValue
    : solvedCount >= state.rules.goalValue;
}

function resetChain(
  state: LevelState,
  action: GameAction,
  extra: Partial<LevelState> = {},
): LevelState {
  return {
    ...state,
    ...extra,
    phase: 'ready',
    chain: { cells: [] },
    preview: null,
    history: appendHistory(state, action),
  };
}

function beginChain(
  state: LevelState,
  action: Extract<GameAction, { type: 'BEGIN_CHAIN' }>,
): LevelState {
  if (state.phase !== 'ready' || !getTile(state.board, action.cell)) return state;
  return {
    ...state,
    phase: 'dragging',
    chain: { cells: [action.cell] },
    preview: null,
    history: appendHistory(state, action),
  };
}

function extendChain(
  state: LevelState,
  action: Extract<GameAction, { type: 'EXTEND_CHAIN' }>,
): LevelState {
  if (state.phase !== 'dragging' && state.phase !== 'previewing') return state;
  const cells = state.chain.cells;
  const previous = cells[cells.length - 1];
  const firstTile = cells[0] ? getTile(state.board, cells[0]) : null;
  const nextTile = getTile(state.board, action.cell);
  if (
    !previous ||
    !firstTile ||
    !nextTile ||
    nextTile.colour !== firstTile.colour ||
    cells.some((cell) => cell.row === action.cell.row && cell.col === action.cell.col) ||
    !areAdjacent(previous, action.cell, state.band.allowDiagonals) ||
    cells.length >= state.band.maxChain
  ) {
    return state;
  }
  const nextCells = [...cells, action.cell];
  const preview =
    nextCells.length >= state.band.minChain
      ? evaluateChain(state.board, nextCells, state.band)
      : null;
  return {
    ...state,
    phase: preview ? 'previewing' : 'dragging',
    chain: { cells: nextCells },
    preview,
    history: appendHistory(state, action),
  };
}

function retractChain(
  state: LevelState,
  action: Extract<GameAction, { type: 'RETRACT_CHAIN' }>,
): LevelState {
  if (
    (state.phase !== 'dragging' && state.phase !== 'previewing') ||
    state.chain.cells.length <= 1
  ) {
    return state;
  }
  const cells = state.chain.cells.slice(0, -1);
  const preview =
    cells.length >= state.band.minChain ? evaluateChain(state.board, cells, state.band) : null;
  return {
    ...state,
    phase: preview ? 'previewing' : 'dragging',
    chain: { cells },
    preview,
    history: appendHistory(state, action),
  };
}

function commit(state: LevelState, action: Extract<GameAction, { type: 'COMMIT' }>): LevelState {
  if (state.phase === 'dragging') {
    return resetChain(state, action, { attemptCount: state.attemptCount + 1, combo: 0 });
  }
  if (state.phase !== 'previewing' || !state.preview) return state;
  const attemptCount = state.attemptCount + 1;
  if (!state.preview.isValid || !eq(state.preview.result, state.target)) {
    return resetChain(state, action, { attemptCount, combo: 0 });
  }

  const refilled = refillAfterRemoval(
    state.board,
    state.chain.cells,
    state.band,
    state.rules,
    state.rngState,
  );
  const combo = state.combo + 1;
  const score = state.score + state.chain.cells.length * 10 * combo;
  const solvedCount = state.solvedCount + 1;
  const movesUsed = state.movesUsed + 1;
  const movesRemaining =
    state.movesRemaining === null ? null : Math.max(0, state.movesRemaining - 1);
  const phase: Phase = objectiveComplete(state, score, solvedCount)
    ? 'levelComplete'
    : movesRemaining === 0
      ? 'levelEnded'
      : 'ready';
  return {
    ...state,
    phase,
    board: refilled.board,
    target: refilled.target,
    chain: { cells: [] },
    preview: null,
    score,
    combo,
    movesUsed,
    movesRemaining,
    solvedCount,
    attemptCount,
    rngState: refilled.rngState,
    history: appendHistory(state, action),
  };
}

function swap(state: LevelState, action: Extract<GameAction, { type: 'SWAP' }>): LevelState {
  if (
    state.phase !== 'ready' ||
    !getTile(state.board, action.a) ||
    !getTile(state.board, action.b) ||
    !areAdjacent(action.a, action.b, state.band.allowDiagonals)
  )
    return state;
  const board = swapTiles(state.board, action.a, action.b);
  if (analyseWithBand(board, state.target, state.rules, state.band).isStuck) return state;
  const movesRemaining =
    state.movesRemaining === null ? null : Math.max(0, state.movesRemaining - 1);
  return {
    ...state,
    phase: movesRemaining === 0 ? 'levelEnded' : 'ready',
    board,
    movesUsed: state.movesUsed + 1,
    movesRemaining,
    history: appendHistory(state, action),
  };
}

/**
 * How many boards `equationShuffle` may draw before it settles for what it has.
 *
 * Same budget `generatePackInternal` uses, for the same reason: generous enough
 * that exhausting it means something is wrong with the band, not with luck.
 * Measured over 400 trials per band, the worst seed needed 203 draws (sprout) and
 * nothing came close to the cap.
 */
const SHUFFLE_ATTEMPTS = 512;

/**
 * A replacement board that respects the band, not just the first one generated.
 *
 * `createInitialState` guarantees a board is SOLVABLE — it plants a solution pair —
 * but it does not enforce `band.maxSolutions`. That ceiling is a difficulty
 * control, and unenforced it runs far over: on sprout only 3.3% of generated
 * boards sit at or under the ceiling of 4, against a median of 14. Handing one of
 * those to a child mid-level makes the puzzle markedly easier than its band
 * intends, which is the whole thing the band is for.
 *
 * Levels reached any other way are already validated — `generatePackInternal`
 * retries until `validatePuzzle` passes — so this power-up was the one live path
 * that skipped the check. Rather than enforce the ceiling inside
 * `createInitialState` (which cannot be done at acceptable cost, and would change
 * rng consumption order and so regenerate every existing level — see ADR-0009),
 * the check lives here, at the call site, exactly as the pack generator does it.
 *
 * Deterministic in `state`, so `dispatch` stays a pure reducer [INV-5]: each
 * rejected board hands its `rngState` to the next draw.
 *
 * Fails OPEN. If the budget runs out this returns the last candidate rather than
 * throwing — a reducer must always return a state, and that board is still
 * solvable by construction. A board that is too easy is a worse level; a thrown
 * exception mid-level is a broken game.
 */
function shuffledLevel(state: LevelState): LevelState {
  let candidate = createInitialState(state.rngState, state.rules, state.band);

  for (let attempt = 1; attempt < SHUFFLE_ATTEMPTS; attempt += 1) {
    if (validatePuzzle(candidate.board, candidate.target, state.rules, state.band).valid) break;
    candidate = createInitialState(candidate.rngState, state.rules, state.band);
  }

  return candidate;
}

function usePowerUp(
  state: LevelState,
  action: Extract<GameAction, { type: 'USE_POWER_UP' }>,
): LevelState {
  if (state.phase !== 'ready' || !state.rules.allowedPowerUps.includes(action.id)) return state;
  if (action.id === 'hintLens' || action.id === 'numberLine') {
    return { ...state, hintsUsed: state.hintsUsed + 1, history: appendHistory(state, action) };
  }
  if (action.id === 'equationShuffle') {
    const fresh = shuffledLevel(state);
    return {
      ...state,
      board: fresh.board,
      target: fresh.target,
      rngState: fresh.rngState,
      history: appendHistory(state, action),
    };
  }
  return { ...state, history: appendHistory(state, action) };
}

export function dispatchGame(state: LevelState, action: GameAction): LevelState {
  switch (action.type) {
    case 'BEGIN_CHAIN':
      return beginChain(state, action);
    case 'EXTEND_CHAIN':
      return extendChain(state, action);
    case 'RETRACT_CHAIN':
      return retractChain(state, action);
    case 'CANCEL_CHAIN':
      return state.phase === 'dragging' || state.phase === 'previewing'
        ? resetChain(state, action)
        : state;
    case 'COMMIT':
      return commit(state, action);
    case 'SWAP':
      return swap(state, action);
    case 'REQUEST_HINT':
      return state.phase === 'ready'
        ? { ...state, hintsUsed: state.hintsUsed + 1, history: appendHistory(state, action) }
        : state;
    case 'USE_POWER_UP':
      return usePowerUp(state, action);
    case 'TICK': {
      if (
        !INPUT_PHASES.includes(state.phase) ||
        state.timeRemainingMs === null ||
        action.deltaMs <= 0
      ) {
        return state;
      }
      const timeRemainingMs = Math.max(0, state.timeRemainingMs - action.deltaMs);
      return {
        ...state,
        timeRemainingMs,
        phase: timeRemainingMs === 0 ? 'levelEnded' : state.phase,
        history: appendHistory(state, action),
      };
    }
    case 'PAUSE':
      return INPUT_PHASES.includes(state.phase)
        ? { ...state, phase: 'paused', history: appendHistory(state, action) }
        : state;
    case 'RESUME':
      return state.phase === 'paused'
        ? {
            ...state,
            phase: state.preview
              ? 'previewing'
              : state.chain.cells.length > 0
                ? 'dragging'
                : 'ready',
            history: appendHistory(state, action),
          }
        : state;
    case 'ADVANCE_PHASE':
      return state;
  }
}

File: flutter/packages/game_engine/lib/src/machine.dart
import 'package:poko_game_engine/src/board.dart';
import 'package:poko_game_engine/src/equation.dart';
import 'package:poko_game_engine/src/generator.dart';
import 'package:poko_game_engine/src/num.dart';
import 'package:poko_game_engine/src/refill.dart';
import 'package:poko_game_engine/src/solver.dart';
import 'package:poko_game_engine/src/types.dart';
import 'package:poko_game_engine/src/validator.dart';

const _inputPhases = <Phase>{Phase.ready, Phase.dragging, Phase.previewing};

List<GameAction> _appendHistory(LevelState state, GameAction action) =>
    <GameAction>[...state.history, action];

bool _objectiveComplete(LevelState state, int score, int solvedCount) =>
    state.rules.objective == ObjectiveType.scoreTide
    ? score >= state.rules.goalValue
    : solvedCount >= state.rules.goalValue;

LevelState _copyState(
  LevelState state, {
  Phase? phase,
  Board? board,
  Num? target,
  Chain? chain,
  ({Equation? value})? preview,
  int? score,
  int? combo,
  int? movesUsed,
  ({int? value})? movesRemaining,
  ({int? value})? timeRemainingMs,
  int? solvedCount,
  int? attemptCount,
  int? hintsUsed,
  LevelRules? rules,
  BandConfig? band,
  int? rngState,
  List<GameAction>? history,
}) => LevelState(
  phase: phase ?? state.phase,
  board: board ?? state.board,
  target: target ?? state.target,
  chain: chain ?? state.chain,
  preview: preview == null ? state.preview : preview.value,
  score: score ?? state.score,
  combo: combo ?? state.combo,
  movesUsed: movesUsed ?? state.movesUsed,
  movesRemaining: movesRemaining == null
      ? state.movesRemaining
      : movesRemaining.value,
  timeRemainingMs: timeRemainingMs == null
      ? state.timeRemainingMs
      : timeRemainingMs.value,
  solvedCount: solvedCount ?? state.solvedCount,
  attemptCount: attemptCount ?? state.attemptCount,
  hintsUsed: hintsUsed ?? state.hintsUsed,
  rules: rules ?? state.rules,
  band: band ?? state.band,
  rngState: rngState ?? state.rngState,
  history: history ?? state.history,
);

LevelState _resetChain(
  LevelState state,
  GameAction action, {
  int? attemptCount,
  int? combo,
}) => _copyState(
  state,
  phase: Phase.ready,
  chain: const Chain(cells: <Cell>[]),
  preview: (value: null),
  attemptCount: attemptCount,
  combo: combo,
  history: _appendHistory(state, action),
);

LevelState _beginChain(LevelState state, BeginChain action) {
  if (state.phase != Phase.ready || getTile(state.board, action.cell) == null) {
    return state;
  }
  return _copyState(
    state,
    phase: Phase.dragging,
    chain: Chain(cells: <Cell>[action.cell]),
    preview: (value: null),
    history: _appendHistory(state, action),
  );
}

LevelState _extendChain(LevelState state, ExtendChain action) {
  if (state.phase != Phase.dragging && state.phase != Phase.previewing) {
    return state;
  }
  final cells = state.chain.cells;
  final previous = cells.isEmpty ? null : cells.last;
  final firstTile = cells.isEmpty ? null : getTile(state.board, cells.first);
  final nextTile = getTile(state.board, action.cell);
  if (previous == null ||
      firstTile == null ||
      nextTile == null ||
      nextTile.colour != firstTile.colour ||
      cells.any(
        (cell) => cell.row == action.cell.row && cell.col == action.cell.col,
      ) ||
      !areAdjacent(
        previous,
        action.cell,
        allowDiagonals: state.band.allowDiagonals,
      ) ||
      cells.length >= state.band.maxChain) {
    return state;
  }
  final nextCells = <Cell>[...cells, action.cell];
  final preview = nextCells.length >= state.band.minChain
      ? evaluateChain(state.board, nextCells, state.band)
      : null;
  return _copyState(
    state,
    phase: preview == null ? Phase.dragging : Phase.previewing,
    chain: Chain(cells: nextCells),
    preview: (value: preview),
    history: _appendHistory(state, action),
  );
}

LevelState _retractChain(LevelState state, RetractChain action) {
  if ((state.phase != Phase.dragging && state.phase != Phase.previewing) ||
      state.chain.cells.length <= 1) {
    return state;
  }
  final cells = state.chain.cells.sublist(0, state.chain.cells.length - 1);
  final preview = cells.length >= state.band.minChain
      ? evaluateChain(state.board, cells, state.band)
      : null;
  return _copyState(
    state,
    phase: preview == null ? Phase.dragging : Phase.previewing,
    chain: Chain(cells: cells),
    preview: (value: preview),
    history: _appendHistory(state, action),
  );
}

LevelState _commit(LevelState state, Commit action) {
  if (state.phase == Phase.dragging) {
    return _resetChain(
      state,
      action,
      attemptCount: state.attemptCount + 1,
      combo: 0,
    );
  }
  final preview = state.preview;
  if (state.phase != Phase.previewing || preview == null) {
    return state;
  }
  final attemptCount = state.attemptCount + 1;
  if (!preview.isValid || preview.result != state.target) {
    return _resetChain(state, action, attemptCount: attemptCount, combo: 0);
  }

  final refilled = refillAfterRemoval(
    state.board,
    state.chain.cells,
    state.band,
    state.rules,
    state.rngState,
  );
  final combo = state.combo + 1;
  final score = state.score + state.chain.cells.length * 10 * combo;
  final solvedCount = state.solvedCount + 1;
  final movesUsed = state.movesUsed + 1;
  final movesRemaining = state.movesRemaining == null
      ? null
      : _max(0, state.movesRemaining! - 1);
  final phase = _objectiveComplete(state, score, solvedCount)
      ? Phase.levelComplete
      : movesRemaining == 0
      ? Phase.levelEnded
      : Phase.ready;
  return _copyState(
    state,
    phase: phase,
    board: refilled.board,
    target: refilled.target,
    chain: const Chain(cells: <Cell>[]),
    preview: (value: null),
    score: score,
    combo: combo,
    movesUsed: movesUsed,
    movesRemaining: (value: movesRemaining),
    solvedCount: solvedCount,
    attemptCount: attemptCount,
    rngState: refilled.rngState,
    history: _appendHistory(state, action),
  );
}

LevelState _swap(LevelState state, Swap action) {
  if (state.phase != Phase.ready ||
      getTile(state.board, action.a) == null ||
      getTile(state.board, action.b) == null ||
      !areAdjacent(
        action.a,
        action.b,
        allowDiagonals: state.band.allowDiagonals,
      )) {
    return state;
  }
  final board = swapTiles(state.board, action.a, action.b);
  if (analyseWithBand(board, state.target, state.rules, state.band).isStuck) {
    return state;
  }
  final movesRemaining = state.movesRemaining == null
      ? null
      : _max(0, state.movesRemaining! - 1);
  return _copyState(
    state,
    phase: movesRemaining == 0 ? Phase.levelEnded : Phase.ready,
    board: board,
    movesUsed: state.movesUsed + 1,
    movesRemaining: (value: movesRemaining),
    history: _appendHistory(state, action),
  );
}

const int _shuffleAttempts = 512;

LevelState _shuffledLevel(LevelState state) {
  var candidate = createInitialState(state.rngState, state.rules, state.band);
  for (var attempt = 1; attempt < _shuffleAttempts; attempt += 1) {
    if (validatePuzzle(
      candidate.board,
      candidate.target,
      state.rules,
      state.band,
    ).valid) {
      break;
    }
    candidate = createInitialState(candidate.rngState, state.rules, state.band);
  }
  return candidate;
}

LevelState _usePowerUp(LevelState state, UsePowerUp action) {
  if (state.phase != Phase.ready ||
      !state.rules.allowedPowerUps.contains(action.id)) {
    return state;
  }
  if (action.id == PowerUpId.hintLens || action.id == PowerUpId.numberLine) {
    return _copyState(
      state,
      hintsUsed: state.hintsUsed + 1,
      history: _appendHistory(state, action),
    );
  }
  if (action.id == PowerUpId.equationShuffle) {
    final fresh = _shuffledLevel(state);
    return _copyState(
      state,
      board: fresh.board,
      target: fresh.target,
      rngState: fresh.rngState,
      history: _appendHistory(state, action),
    );
  }
  return _copyState(state, history: _appendHistory(state, action));
}

LevelState dispatchGame(LevelState state, GameAction action) =>
    switch (action) {
      BeginChain() => _beginChain(state, action),
      ExtendChain() => _extendChain(state, action),
      RetractChain() => _retractChain(state, action),
      CancelChain() =>
        state.phase == Phase.dragging || state.phase == Phase.previewing
            ? _resetChain(state, action)
            : state,
      Commit() => _commit(state, action),
      Swap() => _swap(state, action),
      RequestHint() =>
        state.phase == Phase.ready
            ? _copyState(
                state,
                hintsUsed: state.hintsUsed + 1,
                history: _appendHistory(state, action),
              )
            : state,
      UsePowerUp() => _usePowerUp(state, action),
      Tick(:final deltaMs) =>
        !_inputPhases.contains(state.phase) ||
                state.timeRemainingMs == null ||
                deltaMs <= 0
            ? state
            : _tick(state, action),
      Pause() =>
        _inputPhases.contains(state.phase)
            ? _copyState(
                state,
                phase: Phase.paused,
                history: _appendHistory(state, action),
              )
            : state,
      Resume() =>
        state.phase == Phase.paused
            ? _copyState(
                state,
                phase: state.preview != null
                    ? Phase.previewing
                    : state.chain.cells.isNotEmpty
                    ? Phase.dragging
                    : Phase.ready,
                history: _appendHistory(state, action),
              )
            : state,
      AdvancePhase() => state,
    };

LevelState _tick(LevelState state, Tick action) {
  final timeRemainingMs = _max(0, state.timeRemainingMs! - action.deltaMs);
  return _copyState(
    state,
    phase: timeRemainingMs == 0 ? Phase.levelEnded : state.phase,
    timeRemainingMs: (value: timeRemainingMs),
    history: _appendHistory(state, action),
  );
}

int _max(int left, int right) => left > right ? left : right;
